Teradata Architecture and Concepts page added (#95)
Teradata/quickstarts: modules/ROOT/pages/teradata-vantage-engine-architecture-and-concepts.adoc

= Teradata Vantage Engine Architecture and Concepts
:experimental:
:page-author: Krutik Pathak
:page-email: krutik.pathak@teradata.com

> :page-revdate: August 3, 2023

Reviewer: ```suggestion
:page-revdate: August 5th, 2023
```

:description: Teradata Vantage Architecture and Concepts.
:keywords: data warehouses, teradata vantage engine architecture, teradata, vantage, Access Module Processors (AMP), Parsing Engines (PE), Massively Parallel Processing (MPP), Virtual Disk (Vdisks), BYNET
:tabs:
:dir: teradata-vantage-architecture-concepts

== Overview

This article explains the underlying concepts of Teradata Vantage engine architecture. All editions of Vantage, including the Primary Cluster in VantageCloud Lake utilize the same engine.  

> Teradata's architecture is designed around the Massively Parallel Processing (MPP), shared-nothing architecture, which enables high-performance data processing and analytics. The MPP architecture distributes the workload into multiple vprocs or virtual processors that are a set of software processes running on a node under Teradata Parallel Database Extensions (PDE) within the multitasking environment of the operating system. vproc where query processing takes place is commonly referred to as an Access Module Processor (AMP). Each AMP is isolated from other AMPs and processes the queries in parallel allowing Teradata to process large volumes of data rapidly. 

Reviewer 2: ```suggestion
Teradata's architecture is designed around a Massively Parallel Processing (MPP), shared-nothing architecture, which enables high-performance data processing and analytics. The MPP architecture distributes the workload into multiple vprocs or virtual processors. The virtual processor where query processing takes place is commonly referred to as an Access Module Processor (AMP). Each AMP is isolated from other AMPs, and processes the queries in parallel allowing Teradata to process large volumes of data rapidly. 
```

The major architectural components of the Teradata Vantage engine include the Parsing Engines (PEs), BYNET, Access Module Processors (AMPs), and Virtual Disks (Vdisks). Vdisks are assigned to AMPs in enterprise platforms, and to the Primary Cluster in the case of VantageCloud Lake environments. 

> image::{dir}/teradata_architecture_major_components.png[Teradata Vantage Major Architectural Components, width=75%]

Reviewer: Would it be useful to number arrows on the diagram to explain the sequence of operations?

> The major architectural components of the Teradata Vantage engine include the Parsing Engines (PEs), BYNET, Access Module Processors (AMPs), and Virtual Disks (Vdisks). Vdisks are applicable to the AMPs in enterprise platforms or the Primary Cluster of a VantageCloud Lake environment. 

Reviewer 2: ```suggestion
The major architectural components of the Teradata Vantage engine include the Parsing Engines (PEs), BYNET, Access Module Processors (AMPs), and Virtual Disks (Vdisks). Vdisks are assigned AMPs in enterprise platforms, and to the Primary Cluster in the case of VantageCloud Lake environments. 
```

== Teradata Vantage Engine Architecture Components
The Teradata Vantage engine consists of the components below:

=== Parsing Engines (PE)
When a SQL query is run in Teradata, it first reaches the Parsing Engine. The functions of the Parsing Engine are:

* Manage individual user sessions (up to 120). 
* Check if the objects used in the SQL query exist.  
* Check if the user has required privileges against the objects used in the SQL query.  
* Parse and optimize the SQL queries.  
* Prepare the execution plan to execute the SQL query and passes it to the corresponding AMPs. 

> * Receives the response from the AMPs and sends it back to the requesting client. 

Reviewer: Style inconsistency - the previous bullets use base verb form

Author: Noted. Fixed.

> * Prepare the execution plan to execute the SQL query and pass it to the corresponding AMPs. 

Reviewer 2: ```suggestion
* Prepare the execution plan to execute the SQL query and passes it to the corresponding AMPs. 
```

=== BYNET 

> BYNET is a system that enables component communication. The BYNET system provides high-speed bi-directional broadcast, multicast, and point-to-point communication and merge functions. It performs three key functions: coordinating multi-AMP queries, reading data from multiple AMPs, regulating message flow to prevent congestion, and processing platform throughput. These functions of BYNET make Vantage highly scalable and achieve Massively Parallel Processing (MPP) capability.  

Reviewer 2: ```suggestion
BYNET is a system that enables component communication. The BYNET system provides high-speed bi-directional broadcast, multicast, and point-to-point communication and merge functions. It performs three key functions: coordinating multi-AMP queries, reading data from multiple AMPs, regulating message flow to prevent congestion, and processing platform throughput. These functions of BYNET make Vantage highly scalable and enable Massively Parallel Processing (MPP) capabilities.  
```

=== Parallel Database Extension (PDE)
Parallel Database Extension (PDE) is an intermediary software layer positioned between the operating system and the Teradata Vantage database. PDE enables MPP systems to use features such as BYNET and shared disks. It facilitates the parallelism that is responsible for the speed and linear scalability of the Teradata Vantage database.  

=== Access Module Processor (AMP)
AMPs are responsible for data storage and retrieval. Each AMP is associated with its own set of Virtual Disks (Vdisks) where the data is stored, and no other AMP can access that content in line with the shared-nothing architecture. The functions of AMP are:

* Access storage using Vantage’s Block File System Software  
* Lock management 
* Sorting rows 
* Aggregating columns 
* Join processing 
* Output conversion 
* Disk space management 
* Accounting 
* Recovery processing 

[NOTE]
====

> AMPs in Vantage Enterprise, and the Primary Cluster in the case of VantageCloud Lake, store data in a Block File System (BFS) format on Vdisks. AMPs in Compute Clusters and Compute Worker Nodes on Lake do not have BFS, they can only access data in object storage using the Object File System (OFS). 

Reviewer: If we say VantageCloud Lake, then we should use the full name for VantageCloud Enterprise.
Also, why does this callout mention only VantageCloud editions? Why not VantageCore IntelilFlex and VantageCore VMware?

Author: Yeah spent some time today to understand this more. I have added about VantageCloud Enterprise, VantageCore Intelliflex and VantageCore VMware too.

> AMPs in Vantage Enterprise or Primary Cluster of VantageCloud Lake have data in Block File System (BFS) and thus have Vdisks. AMPs in Compute Cluster and Compute Worker Nodes on Lake do not have BFS, they can only access data in object storage using the Object File System (OFS). 

Reviewer 2: ```suggestion
AMPs in Vantage Enterprise, and the Primary Cluster in the case of VantageCloud Lake, store data in a Block File System (BFS) format on Vdisks. AMPs in Compute Clusters and Compute Worker Nodes on Lake do not have BFS, they can only access data in object storage using the Object File System (OFS). 
```

====

=== Virtual Disks (Vdisks)
These are units of storage space owned by an AMP. Virtual Disks are used to hold user data (rows within tables). Virtual Disks map to physical space on a disk.

=== Node
A node, in the context of Teradata systems, represents an individual server that functions as a hardware platform for the database software. It serves as a processing unit where database operations are executed under the control of a single operating system. When Teradata is deployed in a cloud, it follows the same MPP, shared-nothing architecture but the physical nodes are replaced with virtual machines (VMs). 

== Teradata Vantage Architecture Concepts
The concepts below are applicable to Teradata Vantage.

=== Linear Growth and Expandability 
Teradata is a linearly expandable RDBMS. As the workload and data volume increase, adding more hardware resources such as servers or nodes results in a proportional increase in performance and capacity. Linear Scalability allows for increased workload without decreased throughput.  

> image::{dir}/performance_impact_of_adding_components.png[Performance Impact of Adding Components, width=75%]

Reviewer: I don't understand what this diagram tries to say.

Author: This is to show how increase in workload can be adjusted with linear scaling. I believe the text itself is enough to give the idea of the concept and there is no need for this diagram.
